api/provider/aws/aws_test/load_balancer_create_test.go

```diff
@@ -0,0 +1,335 @@
+	mockConfig := mock_config.NewMockAPIConfig(ctrl)
+
+	mockConfig.EXPECT().Instance().Return("test").AnyTimes()
+	mockConfig.EXPECT().VPC().Return("vpc_id").AnyTimes()
+	mockConfig.EXPECT().Region().Return("region").AnyTimes()
+	mockConfig.EXPECT().AccountID().Return("123456789012")
+	mockConfig.EXPECT().PrivateSubnets().Return([]string{"priv1", "priv2"}).AnyTimes()
+	mockConfig.EXPECT().PublicSubnets().Return([]string{"pub1", "pub2"}).AnyTimes()
+
+	defer provider.SetEntityIDGenerator("lb_id")()
+
+	req := models.CreateLoadBalancerRequest{
+		LoadBalancerName: "lb_name",
+		EnvironmentID:    "env_id",
+		IsPublic:         true,
+		Ports: []models.Port{
+			{
+				ContainerPort: 80,
+				HostPort:      80,
+				Protocol:      "tcp",
+			},
+		},
+		HealthCheck: models.HealthCheck{
+			Target:             "TCP:80",
+			Interval:           30,
+			Timeout:            5,
+			HealthyThreshold:   2,
+			UnhealthyThreshold: 2,
+		},
+	}
+
+	readSGHelper(mockAWS, "l0-test-env_id-env", "env_sg")
+
+	createSGHelper(t, mockAWS, "l0-test-lb_id-lb", "vpc_id")
+	readSGHelper(mockAWS, "l0-test-lb_id-lb", "lb_sg")
+
+	ingressInput := &ec2.AuthorizeSecurityGroupIngressInput{}
+	ingressInput.SetGroupId("lb_sg")
+	ingressInput.SetCidrIp("0.0.0.0/0")
+	ingressInput.SetIpProtocol("TCP")
+	ingressInput.SetFromPort(int64(80))
+	ingressInput.SetToPort(int64(80))
+
+	mockAWS.EC2.EXPECT().
+		AuthorizeSecurityGroupIngress(ingressInput).
+		Return(&ec2.AuthorizeSecurityGroupIngressOutput{}, nil)
+
+	iamRoleInput := &iam.CreateRoleInput{}
+	iamRoleInput.SetRoleName("l0-test-lb_id-lb")
+	iamRoleInput.SetAssumeRolePolicyDocument(provider.DEFAULT_ASSUME_ROLE_POLICY)
+
+	mockAWS.IAM.EXPECT().
+		CreateRole(iamRoleInput).
+		Return(&iam.CreateRoleOutput{}, nil)
+
+	renderedLBRolePolicy, err := provider.RenderLoadBalancerRolePolicy(
+		"region",
+		"123456789012",
+		"l0-test-lb_id",
+		provider.DEFAULT_LB_ROLE_POLICY_TEMPLATE)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	putIAMPolicyInput := &iam.PutRolePolicyInput{}
+	putIAMPolicyInput.SetPolicyName("l0-test-lb_id-lb")
+	putIAMPolicyInput.SetRoleName("l0-test-lb_id-lb")
+	putIAMPolicyInput.SetPolicyDocument(renderedLBRolePolicy)
+
+	mockAWS.IAM.EXPECT().
+		PutRolePolicy(putIAMPolicyInput)
+
+	listeners := make([]*elb.Listener, len(req.Ports))
+	listener := &elb.Listener{}
+	listener.SetProtocol("tcp")
+	listener.SetLoadBalancerPort(80)
+	listener.SetInstancePort(80)
+	listener.SetInstanceProtocol("tcp")
+	listeners[0] = listener
+
+	createLoadBalancerInput := &elb.CreateLoadBalancerInput{}
+	createLoadBalancerInput.SetLoadBalancerName("l0-test-lb_id")
+	createLoadBalancerInput.SetScheme("internet-facing")
+	createLoadBalancerInput.SetSecurityGroups([]*string{aws.String("env_sg"), aws.String("lb_sg")})
+	createLoadBalancerInput.SetSubnets([]*string{aws.String("pub1"), aws.String("pub2")})
+	createLoadBalancerInput.SetListeners(listeners)
+
+	mockAWS.ELB.EXPECT().
+		CreateLoadBalancer(createLoadBalancerInput).
```

Comment:
I would use the `mock.EXPECT().Do(validate)` pattern to run assertions on only the listeners here.

Reply:
:white_check_mark: